### PR TITLE
Add github action and adjust all code to hlint/stylish

### PR DESCRIPTION
**Pull request by paweljakubas**
### Issue Number

ADP-2477

### Comments

- [x] Added both `hlint` and `stylish-haskell` GitHub actions, with versions that are compatible with nix build tool versions.
- [x] Adjusted code to comply with `hlint` suggestions and satisfy `stylish-haskell`.

### CODE REVIEW

**Comment by Anviking** [comment]
Looks great! (modulo remaining choice whether to fix hlints + upgrade nix-shell version vs downgrade ci version) 

Minor point: is it not possible to put hlint and stylish-haskell in separate "sections" (whatever notion of section is appropriate) 
Skärmavbild 2022-12-02 kl  15 42 17
so that it is more easily apparent which one of them failed, rather than having one "Check hlint and stylish"?

File: scripts/exe-from-github.sh [hunk 1]
```diff
@@ -19,7 +19,11 @@
 #   PACKAGE              Name of the release package
 #   PLATFORM             Target platform of the executable
 #   VERSION              Executable version number
+<<<<<<< HEAD
```
Comment (Unisay) [inline scripts/exe-from-github.sh]: This doesn't look right

Comment (paweljakubas) [reply scripts/exe-from-github.sh]: indeed, rebase artefact. thanks, fixed

File: lib/wallet/bench/latency-bench.hs [hunk 2]
```diff
@@ -415,8 +414,8 @@ walletApiBench capture ctx = do
         pure ()
      where
         arbitraryStake :: Maybe Coin
-        arbitraryStake = Just $ ada 10000
-            where ada = Coin . (1000*1000*)
+        arbitraryStake = Just $ ada 10_000
+            where ada = Coin . (1_000*1_000*)
```
Comment (jonathanknowles) [inline lib/wallet/bench/latency-bench.hs]:
```suggestion
            where ada = Coin . (1_000_000 *)
```

File: lib/wallet/bench/latency-bench.hs [hunk 3]
```diff
@@ -433,7 +432,7 @@ withShelleyServer
 withShelleyServer tracers action = do
     ctx <- newEmptyMVar
     let setupContext np baseUrl = do
-            let sixtySeconds = 60*1000*1000 -- 60s in microseconds
+            let sixtySeconds = 60*1_000*1_000 -- 60s in microseconds
```
Comment (jonathanknowles) [inline lib/wallet/bench/latency-bench.hs]:
```suggestion
            let sixtySeconds = 60 * 1_000_000 -- 60s in microseconds
```

File: lib/wallet/integration/src/Test/Integration/Framework/DSL.hs [hunk 4]
```diff
@@ -2144,7 +2144,7 @@ json = aesonQQ
 
 arbitraryStake :: Maybe Coin
 arbitraryStake = Just $ ada 10_000_000_000
-  where ada = Coin . (1000*1000*)
+  where ada = Coin . (1_000*1_000*)
```
Comment (jonathanknowles) [inline lib/wallet/integration/src/Test/Integration/Framework/DSL.hs]:
```suggestion
  where ada = Coin . (1_000_000 *)
```

File: lib/wallet/integration/src/Test/Integration/Scenario/API/Shelley/Settings.hs [hunk 5]
```diff
@@ -90,4 +90,4 @@ spec = describe "SHELLEY_SETTINGS" $ do
 
 arbitraryStake :: Maybe Coin
 arbitraryStake = Just $ ada 10_000
-  where ada = Coin . (1000*1000*)
+  where ada = Coin . (1_000*1_000*)
```
Comment (jonathanknowles) [inline lib/wallet/integration/src/Test/Integration/Scenario/API/Shelley/Settings.hs]:
```suggestion
  where ada = Coin . (1_000_000 *)
```

File: lib/wallet/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs [hunk 6]
```diff
@@ -1122,11 +1122,10 @@ instance (Arbitrary a, Ord a) => Arbitrary (NonSingletonRange a) where
     arbitrary = do
         -- Iterate through the infinite list of arbitrary ranges and return
         -- the first range that is not a singleton range:
-        ranges <- infiniteList
-        pure $ head $ catMaybes $
-            makeNonSingletonRangeValid . NonSingletonRange <$> ranges
-    shrink (NonSingletonRange r) = catMaybes $
-        makeNonSingletonRangeValid . NonSingletonRange <$> shrink r
+        head . mapMaybe (makeNonSingletonRangeValid . NonSingletonRange)
+               <$> infiniteList
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs]:
```suggestion
        head . mapMaybe (makeNonSingletonRangeValid . NonSingletonRange)
            <$> infiniteList
```

File: .github/workflows/style.yml [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    tags: [ "*.*.*" ]
+
+jobs:
+  build:
+    strategy:
+      matrix:
+        os: [ubuntu-20.04]
+    runs-on: ${{ matrix.os }}
+    steps:
+    - name: 📥 Checkout repository
```
Comment (jonathanknowles) [inline .github/workflows/style.yml]:
```suggestion
    - name: 📥 Check out repository
```

File: prototypes/light-mode-test/src/Light/ReadBlocks.hs [hunk 8]
```diff
@@ -137,15 +136,13 @@ lightSync follower = do
...
-                old <- rollBackward follower $ fromBlock target
-                nextBlockFrom old
-
-    stabilityWindow = 1000 :: Integer
+        (if b then
+             (do target <- BF.getBlock . Left $ height - stabilityWindow
+                 old <- rollBackward follower $ fromBlock target
+                 nextBlockFrom old)
+         else
+             fromBlock <$> BF.getLatestBlock)
```
Comment (jonathanknowles) [inline prototypes/light-mode-test/src/Light/ReadBlocks.hs]: I think the multi-line parenthesis is much harder to read than the original block layout.

Can we keep the old pattern match here?
```suggestion
        case b of
            False -> fromBlock <$> BF.getLatestBlock
            True -> do
                -- fall back to behind the stability window
                target <- BF.getBlock . Left $ height - stabilityWindow
                old <- rollBackward follower $ fromBlock target
                nextBlockFrom old
```

**Review by jonathanknowles — CHANGES_REQUESTED** [review]
Hi @paweljakubas 

Thanks for adding this GitHub action!

I do wonder about some of the changes though. Would it not have been possible to add some "surgical" `HLINT` statements to tops of files to avoid having to create such a large PR?

I like the use of `NumericUnderscores` though, as you know!

Two things I would like to see:
1. Could you link the PR to a JIRA ticket? (So we can track the work we're doing to rebuild our CI system.)
2. Could you fill out the PR description, to briefly state what you've done, and why?
3. Are you able to arrange these commits by **_type_** of change? For example:
    - Commit 1: `Use NumericUnderscores.`
    - Commit 2: `Use void instead of "() <$".`
    - Commit 3: `Apply eta reductions.`
 
 I would say that it's much easier to review if we can see a single commit per type of change, rather than separating commits arbitrarily by area of the code, as there's less context switching for reviewers (and people reading the commit history).
 
Alternatively, if separating commits in this way is too much work, then I would propose squashing the commits that make changes to the code, as there's not much benefit to separating them out as they are now.

File: .github/workflows/style.yml (hunk 7 above)
Comment (paweljakubas) [reply .github/workflows/style.yml]: done

File: prototypes/light-mode-test/src/Light/ReadBlocks.hs (hunk 8 above)
Comment (paweljakubas) [reply prototypes/light-mode-test/src/Light/ReadBlocks.hs]: sounds reasonable

**Comment by paweljakubas** [comment]
Thanks @jonathanknowles for your time.  Couple things from me:
1. I have created the ticket in Jira
2. Applied all your suggestions
3. Now Coding Standards GREEN
4. The changes (mainly hlint) were needed as previously we had many holes in coverage! Now I cover EVERYTHING in lib and prototypes library. and I do think it is the right thing to do. Making numerous exceptions is against those coverage drive I believe in. Due to that enlarged coverage so many changes.
5. Although I think it is valid to decompose PR and organize it I would strongly suggest not to do it here. I spent already two days on this, followed diligently what hlint suggested to make it green. And I think most changes suggested are reasonable. I would appreciate if you were not so demanding and strict here (although I revere your level of detail and practices vehemently) ;-)

@Anviking we are green!

**Review by jonathanknowles — APPROVED** [review]
Hi @paweljakubas 

Many thanks for filling out the PR description and linking to the JIRA ticket (https://input-output.atlassian.net/browse/ADP-2477), I appreciate it! 🙇🏻 

I took the liberty of assigning https://input-output.atlassian.net/browse/ADP-2477 to the [current sprint](https://input-output.atlassian.net/jira/software/c/projects/ADP/boards/231). Hope that's okay with you! My reason for doing this is so that we can track this item of work and account for it together with all the other pieces of work that our team has had to pick up in the wake of our old CI system going away. I think it's important that your work is noticed and recognised!

As for the commit structure, I think it's fine, and understand that it's too painful to reorganise at this stage. Though perhaps squashing some of the commits together might leave a cleaner history?

Thanks again for your hard work in making this PR. 💯

File: lib/coin-selection/test/spec/Cardano/CoinSelection/CollateralSpec.hs [hunk 9]
```diff
@@ -937,11 +937,11 @@ unitTests_numberOfSubsequencesOfSize_withinBounds = unitTests
...
-        [ (100,  1,                 100)
-        , (100,  2,                4950)
-        , (100,  4,             3921225)
-        , (100,  8,        186087894300)
-        , (100, 16, 1345860629046814650)
+        [ (100,  1,                       100)
+        , (100,  2,                     4_950)
+        , (100,  4,                 3_921_225)
+        , (100,  8,           186_087_894_300)
+        , (100, 16, 1_345_860_629_046_814_650)
```
Comment (jonathanknowles) [inline lib/coin-selection/test/spec/Cardano/CoinSelection/CollateralSpec.hs:944]: 👍🏻

**Review by Anviking — APPROVED** [review]
🎉 

Just, with the .hs changes, could we wait for #3627 and see buildkite green for added re-assurance? I _think_ the docker check is compiling the code, at least to some extent.

**Comment by Anviking** [comment]
> Just, with the .hs changes, could we wait for https://github.com/input-output-hk/cardano-wallet/pull/3627 and see buildkite green for added re-assurance? I think the docker check is compiling the code, at least to some extent.

Update: @paweljakubas, #3627 is merged, feel free to rebase and merge

**Review by paolino — APPROVED** [review]
This is very nice! Thanks for syntax-shaking the tree! ... Fourmolu ... I can hear you coming!

**Comment by Anviking** [comment]
Rebased to get #3631 included, as buildkite didn't seem to be running

File: .github/workflows/style.yml [hunk 10]
```diff
@@ -0,0 +1,49 @@
+name: Coding Standards
+
+on: [push]
+
+jobs:
+  build:
+    runs-on: self-hosted
```
Comment (piotr-iohk) [inline .github/workflows/style.yml]: why self-hosted?

File: .github/workflows/style.yml [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    - name: 🔬 Check hlint
+      shell: bash
+      env:
+        PACKAGE: hlint
+        REPOSITORY: ndmitchell/hlint
+        PLATFORM: x86_64-linux
+        RELEASE: v3.3.1
+        VERSION: 3.3.1
+      run: |
+        apt install -y libtinfo5
```
Comment (piotr-iohk) [inline .github/workflows/style.yml]: seem redundant since there is the same command on "📥 Add packages"?

**Pull request merged**